src/TLuaInterpreter.cpp

```diff
@@ -9865,6 +9865,14 @@ int TLuaInterpreter::appendCmdLine( lua_State * L )
     return 0;
 }
 
+int TLuaInterpreter::getCmdLine( lua_State * L )
+{
+    Host * pHost = TLuaInterpreter::luaInterpreterMap[L];
+    QString curText = pHost->mpConsole->mpCommandLine->toPlainText();
+    lua_pushstring( L, curText.toLatin1().data() );
```

Comment:
Yep, toLatin1 will barf on anything that isn't strictly ASCII and Utf-8 is
the way to go, into and out of the Lua sub-system, and given that many of
the strings that get passed into that system are const QStrings that
shouldn't be modified by the called function it is safest to use
constData() rather than data() so that modifications are done on copies.

Try and use tr() to declare UI strings, (and QStringLiteral()) for ones the
user doesn't see - that will make later i18n and l10n work much easier with
the Qt translation system.  There is a sort of Lua error message format
that I've been using:

For input type problems causing fatal errors:
```
lua_pushstring( L, tr( "<function name without ()>: bad argument #%1 type
(<argument "name"> as <string|number|boolean|nil(!?)> expected, got %2!)"
).arg( <arg number> ).arg( lual_gettype( L , <arg number> )
).toUtf8().constData() );
lua_error( L )
return 1;
```
though for most cases we put the argument number directly in the "format"
string and thus lose the first ".arg(...)" off the tr() call.

There are some cases where the argument can be more than one type (e.g.
room name or room id number) and other cases where the argument is optional
and may be omitted so the wording changes.

For input problems causing no output say because the input is out of range:
```
lua_pushnil( L );
lua_pushstring( L, tr( "<function name without ()>: bad argument #%1 value
(<argument "name"> {must be as...|is not a valid...|cannot be empty string|
etc. }, got <\" - for strings>%2<\" - for strings>.)" ).arg( <arg number>
).arg( lual_gettype( <entried value if displayable> )
).toUtf8().constData() );
return 2;
```

There are oddities which make things more complicated and it is usually
wise to check for a valid Host * pHost and, for map related stuff, for a
valid TMap * or TRoomDB * in case the user HASN'T got a map or mapper
loaded/active.

In processing string data coming from the Lua subsystem use QStrings if you
can and convert them with:
QString <variableName> = QString::fromUtf8( lua_tostring(L, <argument
number>) );

You will find plenty of cases in TLuaInterpreter where a:
```
std::string <variable> = lua_tostring(L , <argument number>)
```
followed by a later:
```
QString _<variableName> = <variableName>.c_str();
```
is used, but I'm not sure there is any benefit now for going via a
std:;string intermediate.


On 4 March 2015 at 10:08, keneanung <notifications@github.com> wrote:

> In src/TLuaInterpreter.cpp
> <https://github.com/Mudlet/Mudlet/pull/251#discussion_r25763612>:
>
> > @@ -9865,6 +9865,14 @@ int TLuaInterpreter::appendCmdLine( lua_State * L )
> >      return 0;
> >  }
> >
> > +int TLuaInterpreter::getCmdLine( lua_State * L )
> > +{
> > +    Host * pHost = TLuaInterpreter::luaInterpreterMap[L];
> > +    QString curText = pHost->mpConsole->mpCommandLine->toPlainText();
> > +    lua_pushstring( L, curText.toLatin1().data() );
>
> [image: :+1:] will change that when I get home today.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/251/files#r25763612>.
>